app/com/arpnetworking/metrics/portal/scheduling/JobExecutorActor.java

```diff
@@ -253,18 +257,22 @@ private void tick(final Tick message) throws ActorNotInitializedException {
             return;
         }
 
-        if (_clock.instant().isBefore(nextRun.get().minus(EXECUTION_SLOP))) {
-            scheduleTickFor(nextRun.get());
+        if (_clock.instant().isBefore(_nextRun.get().minus(EXECUTION_SLOP))) {
+            scheduleTickFor(_nextRun.get());
         } else {
             try {
-                attemptExecuteAndUpdateRepository(nextRun.get());
+                attemptExecuteAndUpdateRepository(_nextRun.get());
             } catch (final NoSuchJobException error) {
                 LOGGER.warn()
                         .setMessage("attempted to start executing job, but job no longer exists in repository")
                         .addData("ref", cachedJob.getRef())
-                        .addData("scheduled", nextRun.get())
+                        .addData("scheduled", _nextRun.get())
                         .log();
                 killSelf();
+            } finally {
+                // This run was attempted so we should clear it and recompute
+                // the scheduled time on the next tick.
+                _nextRun = Optional.empty();
```

Comment:
This is the only thing I'm unsure about - when to reset the cached value of nextRun. IMO it should either be here or within `JobCompleted` but I've placed it here in case the Job doesn't actually complete successfully. I don't think it should make much of a difference because in the case the job did complete successfully, the actor will have to process that message before any tick it receives.

Reply:
There is a slight inconsistency with clearing it. Based on the comments cleared implies `// First tick since clearing next run, refresh.`. Should this path actually reset it instead of clearing it?

Also please verify that `attemptExecuteAndUpdateRepository` is synchronous.

Reply:
I think you're correct, I think refreshing is the better choice here.

`attemptExecuteAndUpdateRepository` is actually asynchronous, it executes the job and pipes the result back to the actor which then marks the job as completed.

I think if I just refresh here, there shouldn't be any issues. It would then be acting the same as before this diff. I would also be concerned that refreshing instead within JobCompleted could make the executor get stuck since if somehow the job were faulty and `attemptExecuteAndUpdateRepository` failed, it would not be able to advance `nextRun`.